modules/openapi-generator/src/main/java/org/openapitools/codegen/utils/ModelUtils.java

```diff
@@ -988,19 +1029,34 @@ public static String getParentName(ComposedSchema composedSchema, Map<String, Sc
                         return parentName;
                     } else {
                         // not a parent since discriminator.propertyName is not set
+                        hasAmbiguousParents = true;
                         refedWithoutDiscriminator.add(parentName);
                     }
                 } else {
-                    // not a ref, doing nothing
+                    // not a ref, doing nothing, except counting the number of times the 'null' type
+                    // is listed as composed element.
+                    if (ModelUtils.isNullType(schema)) {
+                        // If there are two interfaces, and one of them is the 'null' type,
+                        // then the parent is obvious and there is no need to warn about specifying
+                        // a determinator.
+                        nullSchemaChildrenCount++;
+                    }
                 }
             }
+            if (refedWithoutDiscriminator.size() == 1 && nullSchemaChildrenCount == 1) {
```

Comment:
Just a heads up that I don't think this is intended to always indicate polymorphism in the specification. An allOf ref with discriminator is _always_ supposed to indicate polymporphism, but the opposite isn't clear. The case that you've defined here, and that I'm assuming was intended to match the `refedWithoutDiscriminator.size() == 1` condition below is a "legacy" implementation from when we'd began implementing `allOf` support but it wasn't fully supported by the generator.

Details:

* [discriminator](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#fixed-fields-20) document starts with "Adds support for polymorphism."
* [Composition and Inheritance (Polymorphism)](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#composition-and-inheritance-polymorphism) states: "To support polymorphism, the OpenAPI Specification adds the discriminator field. When used, the discriminator will be the name of the property that decides which schema definition validates the structure of the model."

It doesn't explain this well enough, but from these two it seems like they're defining subtype polymorphism since other types of polymorphism are either irrelevant or don't make much sense in this usage.

_More discussion (we could turn it into an issue as well):_

There are additional details of OpenAPI Spec which aren't explained very well with `allOf`. For example, the discriminator for allOf can be in the `$ref` schema and _not_ in the schema containing the `allOf` definition. In this case, the discriminator will use the schema name of the schema with the `allOf` and $ref. Example:

```
Pet:
  …
  discriminator:
    propertyName: type
  properties:
    name: type
    type: String
Cat:
  allOf:
    - $ref: #/components/schemas/Pet
```

Even without the discriminator, I don't think the `LOGGER.warn` message is totally accurate. For example, the json-schema [combining](https://json-schema.org/understanding-json-schema/reference/combining.html#allof) doc shows this example:

```
{
  "definitions": {
    "address": {
      "type": "object",
      "properties": {
        "street_address": { "type": "string" },
        "city":           { "type": "string" },
        "state":          { "type": "string" }
      },
      "required": ["street_address", "city", "state"]
    }
  },

  "allOf": [
    { "$ref": "#/definitions/address" },
    { "properties": {
        "type": { "enum": [ "residential", "business" ] }
      }
    }
  ]
}
```

Unfortunately, this doesn't define whether a codegen should model:

```
Address:
  + StreetAddress: String
  + City: String
  + State: String

TypedAddress <Address>:
    Enum: [ Residential, Business ]
```

or:

```
Address:
  + StreetAddress: String
  + City: String
  + State: String
    Enum: [ Residential, Business ]
```

In this issue, there's no clarification: https://github.com/OAI/OpenAPI-Specification/issues/1467

In fact, if you consider from a REST input modeling perspective, it might make sense that all allOf without discriminator are unique types with properties of the `$ref`.

I wonder if it may make sense for us to support a vendor extension as a hint whether the user wants to use composition or inheritance for these one-off allOf without a discriminator. Something like:

* `x-codegen-prefer-polymorphism: true`. When no discriminator is defined, we'll _try_ to create a parent and child relationship if the target language/generator supports it
* `x-codegen-prefer-polymorphism: false`. When no discriminator is defined, we'll merge all properties into one "super" class.

The latter might be challenging to support the spec requirement that these are validated independently like in the case where two refs have the same property with different validation characteristics. This isn't a case we handle now.

Reply:
Thank you for the quick feedback. Let me digest this and get back to you tomorrow.

Reply:
> 
> Unfortunately, this doesn't define whether a codegen should model:
> 
> ```
> Address:
>   + StreetAddress: String
>   + City: String
>   + State: String
> 
> TypedAddress <Address>:
>     Enum: [ Residential, Business ]
> ```
> 
> or:
> 
> ```
> Address:
>   + StreetAddress: String
>   + City: String
>   + State: String
>     Enum: [ Residential, Business ]
> ```
> 

Above did you mean? 1) the generated code; 2) the serialization format of the JSON document; or 3) both? If (1), I agree it's not defined in the OAS spec, but OAS does not mandate how code generators should be implemented. If (2), isn't https://tools.ietf.org/html/draft-handrews-json-schema-02  section 9.2.1.1 mandating that it must be the second case?

Reply:
> * `x-codegen-prefer-polymorphism: true`. When no discriminator is defined, we'll _try_ to create a parent and child relationship if the target language/generator supports it
> * `x-codegen-prefer-polymorphism: false`. When no discriminator is defined, we'll merge all properties into one "super" class.
> 
> The latter might be challenging to support the spec requirement that these are validated independently like in the case where two refs have the same property with different validation characteristics. This isn't a case we handle now.

I see there are related discussions about defined "subclassOf", but it does not seem to make much progress.

I agree with all your points, they will have to be addressed in separate PRs. For example, I had an issue with inlined schema exactly as you mentioned, and I ended up doing a workaround by using a $ref.

Reply:
Oh. I wasn't suggesting any additional changes here. I just wanted to comment on the allOf change because of the ambiguity, and hoped it'd generate discussion. There's a lot to be done for allOf support, and I'm hoping we'll make great progress in the 5.0 release.

Regarding your comment:

> If (2), isn't https://tools.ietf.org/html/draft-handrews-json-schema-02 section 9.2.1.1 mandating that it must be the second case?

The issue is that JSON Schema focuses on validation only and not on structure. This is further complicated because OpenAPI Specification disassociates it's `allOf` from JSON Schema:

> allOf - Inline or referenced schema MUST be of a Schema Object and not a standard JSON Schema.

I think the only sane way to support it cleanly is by having good defaults and vendor extensions to modify each model's behavior. We may also want to consider a configuration option at generation to define a "global" default.